### PR TITLE
[PLAT-851] Storage internationalization frontend

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

This allows the user to interact with their storage regions in various ways.

## Changes

- show storage location on node settings page
screen shot 2018-05-18 at 4 07 26 pm
- Have fangorn show region names on provider listings
screen shot 2018-05-18 at 4 09 11 pm
- allow user to set storage location on node creation 
screen shot 2018-05-18 at 4 10 50 pm
- allow user to change thier default node location

user_profile_storage_location

- Update welcome email

## QA Notes

- check welcome email text is correct.
- check fangorn file tree storage info is correct
- check/change the default user storage location
- check/change the default when storage location when creating a node
- check the node setting page to confirm the storage location.

## Documentation

Should be covered in the osf.help docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/projects/PLAT/issues/PLAT-851

### CODE REVIEW

File: website/templates/emails/welcome.html.mako [hunk 1]
```diff
@@ -11,35 +11,46 @@
 
 Hello ${user.fullname},<br>
 <br>
-Thank you for verifying your OSF account, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:<br>
+Thank you for verifying your on OSF, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:
```
Comment (caseyrollins) [inline website/templates/emails/welcome.html.mako]: "Thank you for verifying your on OSF"
should be 
"Thank you for verifying your account on OSF"

File: website/templates/emails/welcome_osf4i.html.mako [hunk 2]
```diff
@@ -11,37 +11,50 @@
 
 Hello ${user.fullname},<br>
 <br>
-Thank you for verifying your OSF account, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:<br>
+Thank you for verifying your on OSF, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:<br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome_osf4i.html.mako]: "Thank you for verifying your on OSF"
should be 
"Thank you for verifying your account on OSF"

File: website/templates/emails/welcome.html.mako (hunk 1 above)
Comment (caseyrollins) [reply website/templates/emails/welcome.html.mako]: "Here are a few things you can do with the OSF"
should be 
"Here are a few things you can do with OSF"

File: website/templates/emails/welcome_osf4i.html.mako (hunk 2 above)
Comment (caseyrollins) [reply website/templates/emails/welcome_osf4i.html.mako]: "Here are a few things you can do with the OSF"
should be 
"Here are a few things you can do with OSF"

File: website/templates/emails/welcome.html.mako [hunk 3]
```diff
@@ -11,35 +11,46 @@
...
 <br>
+
+<h4>Select your storage location</h4>
+Files can be stored in a location you specify from the available geographic regions for new projects. <a href="https://osf.io/settings/account/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Set storage location.</a><br>
+<br>
+
 <h4>Store your files</h4>
-Archive your materials, data, manuscripts, or anything else associated with your research during the research process or after it is complete.<br>
+Archive your materials, data, manuscripts, or anything else associated with your work research during the research process or after it is complete. <a href="http://help.osf.io/m/projectfiles/l/482002-upload-files/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Learn how.</a><br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome.html.mako]: "Archive your materials, data, manuscripts, or anything else associated with your work research"
should be 
"Archive your materials, data, manuscripts, or anything else associated with your work"

File: website/templates/emails/welcome.html.mako [hunk 4]
```diff
@@ -11,35 +11,46 @@
 
 Hello ${user.fullname},<br>
 <br>
-Thank you for verifying your OSF account, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:<br>
+Thank you for verifying your on OSF, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:
 <br>
+
+<h4>Select your storage location</h4>
+Files can be stored in a location you specify from the available geographic regions for new projects. <a href="https://osf.io/settings/account/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Set storage location.</a><br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome.html.mako]: Can we update the link to anchor to the part of the settings page where you set your default region?

File: website/templates/emails/welcome.html.mako [hunk 5]
```diff
@@ -11,35 +11,46 @@
...
 <br>
+
 <h4>Measure your impact</h4>
-You can monitor traffic to your public projects and downloads of your public files.<br>
+You can monitor traffic to your public projects and downloads of your public files. <a href="http://help.osf.io/m/projects/l/524052-view-analytics/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Discover analytics.</a><br>
 <br>
+
 <h4>Connect services that you use</h4>
-GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, Zotero.  Do you use any of these? Link the services that you use to your OSF projects so that all parts of your project are in one place.<br>
+OSF integrates with GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, and Zotero. Do you use any of these? Link the services that you use to your OSF projects so that all parts of your research project are in one place. <a href="http://help.osf.io/m/addons/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Learn about add-ons.</a><br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome.html.mako]: "OSF integrates with GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, and Zotero. Do you use any of these? Link the services that you use to your OSF projects so that all parts of your research project are in one place"
should be 
"OSF integrates with GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, and Zotero. Link the services that you use to your OSF projects so that all parts of your research are in one place"

File: website/templates/emails/welcome.html.mako [hunk 6]
```diff
@@ -11,35 +11,46 @@
...
 <br>
+
 <h4>Collaborate</h4>
-Add your collaborators to have a shared environment for maintaining your research materials and data and never lose files again.<br>
+Add your collaborators to have a shared environment for maintaining your research materials and data and never lose files again. <a href="http://help.osf.io/m/collaborating/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Start collaborating.</a><br>
 <br>
-<br>
-Learn more about the OSF by reading the <a href="http://help.osf.io/">Guides</a>, or email <a href="mailto:${osf_contact_email}">${osf_contact_email}</a> with questions for support.<br>
+
+Learn more about the OSF by reading the <a href="http://help.osf.io/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Guides</a>, or email <a href="mailto:${osf_support_email}">${osf_support_email}</a> for support.<br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome.html.mako]: "Learn more about the OSF by reading the"
should be 
"Learn more about OSF by reading the"

File: website/templates/emails/welcome_osf4i.html.mako [hunk 7]
```diff
@@ -11,37 +11,50 @@
...
 <br>
+
 <h4>Measure your impact</h4>
-You can monitor traffic to your public projects and downloads of your public files.<br>
+You can monitor traffic to your public projects and downloads of your public files. <a href="http://help.osf.io/m/projects/l/524052-view-analytics/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Discover analytics.</a><br>
 <br>
+
 <h4>Connect services that you use</h4>
-GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, Zotero.  Do you use any of these? Link the services that you use to your OSF projects so that all parts of your project are in one place.<br>
+OSF integrates with GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, and Zotero. Do you use any of these? Link the services that you use to your OSF projects so that all parts of your research project are in one place. <a href="http://help.osf.io/m/addons/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Learn about add-ons.</a><br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome_osf4i.html.mako]: "OSF integrates with GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, and Zotero. Do you use any of these? Link the services that you use to your OSF projects so that all parts of your research project are in one place"
should be 
"OSF integrates with GitHub, Dropbox, Google Drive, Box, Dataverse, figshare, Amazon S3, ownCloud, Bitbucket, GitLab, OneDrive, Mendeley, and Zotero. Link the services that you use to your OSF projects so that all parts of your research are in one place"

File: website/templates/emails/welcome_osf4i.html.mako [hunk 8]
```diff
@@ -11,37 +11,50 @@
...
 <br>
+
 <h4>Collaborate</h4>
-Add your collaborators to have a shared environment for maintaining your research materials and data and never lose files again.<br>
+Add your collaborators to have a shared environment for maintaining your research materials and data and never lose files again. <a href="http://help.osf.io/m/collaborating/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Start collaborating.</a><br>
 <br>
 <br>
-Learn more about the OSF at our <a href="http://help.osf.io">Guides page</a>, or email <a href="mailto:${osf_contact_email}">${osf_contact_email}</a> with questions for support.<br>
+Learn more about the OSF by reading the <a href="http://help.osf.io/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Guides</a>, or email <a href="mailto:${osf_support_email}">${osf_support_email}</a> for support.<br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome_osf4i.html.mako]: "Learn more about the OSF by reading the"
should be 
"Learn more about OSF by reading the"

File: website/templates/emails/welcome_osf4i.html.mako [hunk 9]
```diff
@@ -11,37 +11,50 @@
...
+Thank you for verifying your on OSF, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:<br>
 <br>
+
+<h4>Select your storage location</h4>
+Files can be stored in a location you specify from the available geographic regions for new projects. <a href="https://osf.io/settings/account/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Set storage location.</a><br>
+<br>
+
 <h4>Store your files</h4>
-Archive your materials, data, manuscripts, or anything else associated with your research during the research process or after it is complete.<br>
+Archive your materials, data, manuscripts, or anything else associated with your work research during the research process or after it is complete. <a href="http://help.osf.io/m/projectfiles/l/482002-upload-files/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Learn how.</a><br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome_osf4i.html.mako]: "Archive your materials, data, manuscripts, or anything else associated with your work research"
should be 
"Archive your materials, data, manuscripts, or anything else associated with your work"

File: website/templates/emails/welcome_osf4i.html.mako [hunk 10]
```diff
@@ -11,37 +11,50 @@
 
 Hello ${user.fullname},<br>
 <br>
-Thank you for verifying your OSF account, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:<br>
+Thank you for verifying your on OSF, a free, open source service maintained by the Center for Open Science. Here are a few things you can do with the OSF:<br>
 <br>
+
+<h4>Select your storage location</h4>
+Files can be stored in a location you specify from the available geographic regions for new projects. <a href="https://osf.io/settings/account/?utm_source=notification&utm_medium=email&utm_campaign=welcome">Set storage location.</a><br>
```
Comment (caseyrollins) [inline website/templates/emails/welcome_osf4i.html.mako]: Can we anchor here?

File: website/routes.py [hunk 11]
```diff
@@ -74,6 +75,17 @@ def get_globals():
             request_login_url = request.url.replace(request.host_url, settings.DOMAIN)
     else:
         request_login_url = request.url
+
+    if user:
+        default_storage_region = user.get_addon('osfstorage').default_region
+        default_storage_region = {'name': default_storage_region.name, 'id': default_storage_region._id}
+        region_list = [{'name': region.name, 'id': region._id} for region in Region.objects.all()]
+        region_list.insert(0, region_list.pop(region_list.index(default_storage_region)))  # default should be at top of list for UI.
+        print(region_list)
```
Comment (caseyrollins) [inline website/routes.py]: Rogue print.

File: website/static/js/accountSettings.js [hunk 12]
```diff
@@ -344,6 +348,48 @@ var ExternalIdentityViewModel = oop.defclass({
...
+            function(profile) {
+                this.profile(profile);
+                this.locationSelected(this.profile().default_storage_location());
+            }.bind(this)
+        );
+
+
+    },
+    urls: {
+        'change': '/api/v1/profile/region/'
```
Comment (caseyrollins) [inline website/static/js/accountSettings.js]: Prefer the verb "update" to "change" for PUT requests (see the other `urls` properties in this file).

Comment (caseyrollins) [reply website/static/js/accountSettings.js]: Along the same lines, `UpdateDefaultStorageLocation` may be a better name for this class.

File: website/static/js/accountSettings.js [hunk 13]
```diff
@@ -344,6 +348,48 @@ var ExternalIdentityViewModel = oop.defclass({
...
+    },
+    changeDefaultStorageLocation: function() {
+        var request = $osf.ajaxJSON('PUT', this.urls.change, {'data': {'region_id': this.locationSelected().id}});
+        request.done(function() {
+            $osf.growl('Success', 'You have successfully changed you default storage location to <b>' + this.locationSelected().name + '</b>.', 'success');
```
Comment (caseyrollins) [inline website/static/js/accountSettings.js]: "You have successfully changed you default storage location"
should be
"You have successfully changed your default storage location"

File: website/profile/utils.py [hunk 14]
```diff
@@ -86,10 +87,14 @@ def serialize_user(user, node=None, admin=False, full=False, is_profile=False, i
         else:
             merged_by = None
 
+        default_storage_region = user.get_addon('osfstorage').default_region
+        region_list = [region for region in Region.objects.all().values('_id', 'name')]
```
Comment (caseyrollins) [inline website/profile/utils.py]: Could you use your `get_storage_region_list` function here?

File: website/templates/profile/account.mako [hunk 15]
```diff
@@ -82,6 +82,38 @@
...
+                                    <div class="help-block">
+                                            <p>
+                                                This is the default storage location that will be applied for new projects and components created after setting
+                                                and will not be retroactively applied to the existing projects and components.
+                                            </p>
+                                    </div>
+
+                                </span>
+                            </div>
+                            <button class="btn btn-primary" data-bind="click: $root.updateDefaultStorageLocation.bind($root)">Update Location</button>
```
Comment (caseyrollins) [inline website/templates/profile/account.mako]: In order to be consistent with the rest of this page, let's change the button text to sentence case ("Update location").

File: website/static/js/addProjectPlugin.js [hunk 16]
```diff
@@ -474,4 +482,34 @@ var Select2Template = {
     }
 };
 
+var SelectStorageLocation = {
+    view: function(ctrl, options) {
+        return m('select.p-t-sm', {config: SelectStorageLocation.config(options)},
+            [
+            window.contextVars.storageRegions.map(function(region) {
```
Comment (caseyrollins) [inline website/static/js/addProjectPlugin.js]: Any reason to pull directly from `window.contextVars` instead of defining a prop in the controller? 
```
self.storageLocations = m.prop(window.contextVars.storageRegions)
```
(same question on LL503)

File: website/views.py [hunk 17]
```diff
@@ -393,3 +401,13 @@ def legacy_share_v1_search(**kwargs):
             message_long='Please use v2 of the SHARE search API available at {}api/v2/share/search/creativeworks/_search.'.format(settings.SHARE_URL)
         )
     )
+
+def get_storage_region_list(user):
```
[the remark this answers is not in the document]
Comment (caseyrollins) [reply website/views.py]: Super minor: this function could be simplified to something like:
```
if user:
    ...
    return region_list.insert(...)
return []
```

File: website/static/js/addProjectPlugin.js [hunk 18]
```diff
@@ -38,6 +38,7 @@ var AddProject = {
         self.showMore = m.prop(false);
         self.newProjectName = m.prop('');
         self.newProjectDesc = m.prop('');
+        self.newProjectStorageLocation = m.prop(window.contextVars.storageRegions[0]); // first storage region is default
```
Comment (caseyrollins) [inline website/static/js/addProjectPlugin.js]: When creating a project from the "My Projects" page, this default is correct (the user default). However, when creating a component via the "Add Component" button on the project overview page, the default should be the parent node's region (not the user default).

File: website/views.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply website/views.py]: The falsy casing here is actually more about making this pass the pre-exsiting view tests without having to mock user for each one, there are many that fail unless a NoneType user is allowed.

File: website/profile/utils.py (hunk 14 above)
Comment (Johnetordoff) [reply website/profile/utils.py]: We could, but it wouldn't actually decrease the amount of code because we need the `default_storage_region` defined explicitly here, not just have the it be the first thing on the list.

File: website/static/js/addProjectPlugin.js (hunk 18 above)
Comment (Johnetordoff) [reply website/static/js/addProjectPlugin.js]: 👍

File: website/project/views/node.py [hunk 19]
```diff
@@ -832,6 +833,17 @@ def _view_project(node, auth, primary=False,
             for key, value in settings.NODE_CATEGORY_MAP.iteritems()
         ]
     }
+
+    region_list = get_storage_region_list(user)
+
+    # Default should be at top of list for UI and for the project overview page the default region
+    # for a component is that of the it's parent node.
+    if user:
```
Comment (caseyrollins) [inline website/project/views/node.py]: Why is this check necessary? `user` isn't referenced in this block

File: website/profile/utils.py (hunk 14 above)
Comment (caseyrollins) [reply website/profile/utils.py]: Hmm, could you pass the whole list and just pop the first [here](https://github.com/CenterForOpenScience/osf.io/pull/8386/files#diff-6b92464a8f46e4590fe59579570e096eR155)? Not a big deal either way though.

File: website/project/views/node.py (hunk 19 above)
Comment (caseyrollins) [inline website/project/views/node.py]: Why is this check necessary? `user` isn't referenced in this block

File: website/views.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (caseyrollins) [reply website/views.py]: Would you mind linking to an example so I can better understand?

File: website/project/views/node.py [hunk 20]
```diff
@@ -832,6 +833,17 @@ def _view_project(node, auth, primary=False,
...
+        default_storage_region = {'name': node.osfstorage_region.name, '_id': node.osfstorage_region._id}
```
Comment (caseyrollins) [inline website/project/views/node.py]: The `osfstorage_region` property makes a query to find the node's storage region:
https://github.com/CenterForOpenScience/osf.io/blob/63dc0c1fc73f3c4464dcab22c273387f2bdef9bd/osf/models/node.py#L992-L996

Instead of calling it three times in this function, how about storing the region as a variable so you can access the `name` and `_id` without performing additional queries?

File: api/nodes/views.py [hunk 21]
```diff
@@ -632,6 +632,21 @@ def get_queryset(self):
                 .values_list('child__pk', flat=True)
         return self.get_queryset_from_request().filter(pk__in=node_pks).can_view(auth.user).order_by('-modified')
 
+    def get_serializer_context(self):
+        context = super(NodeChildrenList, self).get_serializer_context()
+        region__id = self.request.query_params.get('region', None)
+        id = None
+        if region__id:
+            try:
+                id = Region.objects.get(_id=region__id).id
```
Comment (sloria) [inline api/nodes/views.py]: Minor: This can be optimized slightly by only querying for the `id` column:

```python
id = Region.objects.filter(_id=region__id).values_list('id', flat=True).get()
````

File: api/nodes/views.py [hunk 22]
```diff
@@ -632,6 +632,21 @@ def get_queryset(self):
...
+            except Region.DoesNotExist:
+                raise InvalidQueryStringError('Region {} is invalid.'.format(region__id))
+
+        context.update({
+            'region_id': id
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/nodes/views.py]: Yes this is necessary because components are created via `NodeChildrenList` not `NodeList`. The original add region back-end ticket forgot about this, so I had to add it. This should probably be a mixin that `NodeChildrenList` not `NodeList` share, but I didn't want to add more backend code to what was already a big PR.

This could have definitely been a "comment your own PR" case.

[the remark this answers is not in the document]
Comment (sloria) [reply api/nodes/views.py]: Ah, I understand now. Thanks for clarifying.

File: .docker-compose.env [hunk 23]
```diff
@@ -17,4 +17,4 @@ BROKER_URL=amqp://guest:guest@192.168.168.167:5672/
 REDIS_URL=redis://192.168.168.167:6379/1
 EMBER_DOMAIN=192.168.168.167
 
-#PYTHONUNBUFFERED=0 # This when set to 0 will allow print statements to be visible in the Docker logs
+PYTHONUNBUFFERED=0 # This when set to 0 will allow print statements to be visible in the Docker logs
```
Comment (sloria) [inline .docker-compose.env]: Why was this changed?

Comment (Johnetordoff) [reply .docker-compose.env]: This is the code equivalent of gum stuck to the bottom of your shoe.

File: api/institutions/authentication.py [hunk 24]
```diff
@@ -108,7 +108,7 @@ def authenticate(self, request):
                 mail=WELCOME_OSF4I,
                 mimetype='html',
                 user=user,
-                osf_contact_email=OSF_CONTACT_EMAIL
+                osf_support_email=OSF_SUPPORT_EMAIL
```
Comment (sloria) [inline api/institutions/authentication.py]: Why was this changed?

File: website/profile/utils.py [hunk 25]
```diff
@@ -86,10 +87,14 @@ def serialize_user(user, node=None, admin=False, full=False, is_profile=False, i
         else:
             merged_by = None
 
+        default_region = user.get_addon('osfstorage').default_region
+        available_regions = [region for region in Region.objects.all().values('_id', 'name')]
```
Comment (sloria) [inline website/profile/utils.py]: Minor: This is equivalent to:

```python
list(Region.objects.all().values('_id', 'name'))
```

File: api/institutions/authentication.py (hunk 24 above)
Comment (sloria) [reply api/institutions/authentication.py]: Oh, I see now. This was part of the desired email change. 👍
